Add some batched gradient tests

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #44505 Batched grad support for: slice, select, diagonal
* **#44494 Add some batched gradient tests**

These tests check (most) operations that are useful for bayesian logistic
regression (BLR) models. Said operators are basically those found in the
log_prob functions of Distributions objects. This PR is not a general,
structured solution for testing batched gradients (see "Alternative
solution" for that), but I wanted to test a small subset of operations
to confirm that the BLR use case works.

There will be follow-up PRs implementing support for some missing
operations for the BLR use case.

Alternative solution
=====================

Ideally, and in the future, I want to autogenerate tests from
common_method_invocations and delete all of the manual tests
introduced by this PR. However, if we were to do this now,
we would need to store the following additional metadata somewhere:
- operator name, supports_batched_grad, allow_vmap_fallback_usage

We could store that metadata as a separate table from
common_method_invocations, or add two columns to
common_method_invocations. Either way that seems like a lot of work and
the situation will get better once vmap supports batched gradients for
all operators (on the fallback path).

I am neutral between performing the alternative approach now v.s. just
manually writing out some tests for these operations, so I picked the
easier approach. Please let me know if you think it would be better to
pursue the alternative approach now.

Test Plan:
- `pytest test/test_vmap.py -v -k "BatchedGrad"`

Differential Revision: [D23650408](https://our.internmc.facebook.com/intern/diff/D23650408)

zou3519 committed:
Add some batched gradient tests

These tests check operations that are useful for bayesian logistic
regression models. Said operators are basically those found in the
log_prob functions of Distributions objects.

Alternative solution
=====================

Ideally, and in the future, I want to autogenerate tests from
common_method_invocations and delete all of the manual tests
introduced by this PR. However, if we were to do this now,
we would need to store the following additional metadata somewhere:
- operator name, supports_batched_grad, allow_vmap_fallback_usage

We could store that metadata as a separate table from
common_method_invocations, or add two columns to
common_method_invocations. Either way that seems like a lot of work and
the situation will get better once vmap supports batched gradients for
all operators (on the fallback path).

I am neutral between performing the alternative approach now v.s. just
manually writing out some tests for these operations, so I picked the
easier approach. Please let me know if you think it would be better to
pursue the alternative approach now.

Test Plan:
- `pytest test/test_vmap.py -v -k "BatchedGrad"`

[ghstack-poisoned]

zou3519 committed:
Update on "Add some batched gradient tests"


These tests check operations that are useful for bayesian logistic
regression (BLR) models. Said operators are basically those found in the
log_prob functions of Distributions objects. This PR is not a general,
structured solution for testing batched gradients (see "Alternative
solution" for that), but I wanted to test a small subset of operations
to confirm that the BLR use case works.

Alternative solution
=====================

Ideally, and in the future, I want to autogenerate tests from
common_method_invocations and delete all of the manual tests
introduced by this PR. However, if we were to do this now,
we would need to store the following additional metadata somewhere:
- operator name, supports_batched_grad, allow_vmap_fallback_usage

We could store that metadata as a separate table from
common_method_invocations, or add two columns to
common_method_invocations. Either way that seems like a lot of work and
the situation will get better once vmap supports batched gradients for
all operators (on the fallback path).

I am neutral between performing the alternative approach now v.s. just
manually writing out some tests for these operations, so I picked the
easier approach. Please let me know if you think it would be better to
pursue the alternative approach now.

Test Plan:
- `pytest test/test_vmap.py -v -k "BatchedGrad"`

[ghstack-poisoned]

zou3519 committed:
Update on "Add some batched gradient tests"


These tests check (most) operations that are useful for bayesian logistic
regression (BLR) models. Said operators are basically those found in the
log_prob functions of Distributions objects. This PR is not a general,
structured solution for testing batched gradients (see "Alternative
solution" for that), but I wanted to test a small subset of operations
to confirm that the BLR use case works.

There will be follow-up PRs implementing support for some missing
operations for the BLR use case.

Alternative solution
=====================

Ideally, and in the future, I want to autogenerate tests from
common_method_invocations and delete all of the manual tests
introduced by this PR. However, if we were to do this now,
we would need to store the following additional metadata somewhere:
- operator name, supports_batched_grad, allow_vmap_fallback_usage

We could store that metadata as a separate table from
common_method_invocations, or add two columns to
common_method_invocations. Either way that seems like a lot of work and
the situation will get better once vmap supports batched gradients for
all operators (on the fallback path).

I am neutral between performing the alternative approach now v.s. just
manually writing out some tests for these operations, so I picked the
easier approach. Please let me know if you think it would be better to
pursue the alternative approach now.

Test Plan:
- `pytest test/test_vmap.py -v -k "BatchedGrad"`

[ghstack-poisoned]

albanD commented on test/test_vmap.py:
```diff
@@ -1455,11 +1455,93 @@ def vector_hessian_product(*vectors):
 
+    def _test_arithmetic(self, op, device, test_grad_grad=True):
+        x = torch.randn(2, 3, requires_grad=True, device=device)
+        y = torch.rand(2, 3, device=device).clamp_min_(0.1).requires_grad_()
```
nit: you could have a `_get_rand_no_zeros()` as you seem to do this quite a lot in this file.

zou3519 replied:
Updated with a _get_rand_no_zeros() funciton

albanD reviewed (approved):
LGTM.

You might want to sync up with @mruberry to see if you can add these to the new Op testing system when you start adding them automatically.

zou3519 committed:
Update on "Add some batched gradient tests"


These tests check (most) operations that are useful for bayesian logistic
regression (BLR) models. Said operators are basically those found in the
log_prob functions of Distributions objects. This PR is not a general,
structured solution for testing batched gradients (see "Alternative
solution" for that), but I wanted to test a small subset of operations
to confirm that the BLR use case works.

There will be follow-up PRs implementing support for some missing
operations for the BLR use case.

Alternative solution
=====================

Ideally, and in the future, I want to autogenerate tests from
common_method_invocations and delete all of the manual tests
introduced by this PR. However, if we were to do this now,
we would need to store the following additional metadata somewhere:
- operator name, supports_batched_grad, allow_vmap_fallback_usage

We could store that metadata as a separate table from
common_method_invocations, or add two columns to
common_method_invocations. Either way that seems like a lot of work and
the situation will get better once vmap supports batched gradients for
all operators (on the fallback path).

I am neutral between performing the alternative approach now v.s. just
manually writing out some tests for these operations, so I picked the
easier approach. Please let me know if you think it would be better to
pursue the alternative approach now.

Test Plan:
- `pytest test/test_vmap.py -v -k "BatchedGrad"`

[ghstack-poisoned]
